google/search/src/server.ts

```diff
@@ -0,0 +1,176 @@
+import { search } from './search.ts'
+import { refineQuery } from './refineQuery.ts'
+import { getNewContext } from './context.ts'
+import * as gptscript from '@gptscript-ai/gptscript'
+import { rmSync } from 'node:fs'
+
+const gptsClient = new gptscript.GPTScript()
+
+const query: string = process.env.QUERY ?? ''
+if (query === '') {
+  console.log('error: no query provided')
+  process.exit(1)
+}
+
+if (process.env.GPTSCRIPT_WORKSPACE_ID === undefined || process.env.GPTSCRIPT_WORKSPACE_DIR === undefined) {
+  console.log('error: GPTScript workspace ID and directory are not set')
+  process.exit(1)
+}
+
+// Simultaneously start the browser and generate our search query.
+const [refinedQuery, context, noJSContext] = await Promise.all([
+  refineQuery(query),
+  getNewContext(path.resolve(process.env.GPTSCRIPT_WORKSPACE_DIR), true),
+  getNewContext(path.resolve(process.env.GPTSCRIPT_WORKSPACE_DIR), false)
+])
+
+// Query Google
+const pageContents = await search(
+  context.context,
+  noJSContext.context,
+  refinedQuery,
+  process.env.MAXRESULTS !== undefined ? parseInt(process.env.MAXRESULTS) : undefined
+)
+
+// Ask gpt-4o-mini to generate an answer
+const tool: gptscript.ToolDef = {
+  agents: [],
+  arguments: { 
+    type: 'object', 
+    properties: {
+      search: {
+        type: 'string',
+        description: 'JSON string containing unfocussed search query and results'
+      }
+    },
+    required: ['search']
+  },
+  chat: false,
+  context: [],
+  credentials: [],
+  description: '',
+  export: [],
+  exportContext: [],
+  globalModelName: '',
+  globalTools: [],
+  jsonResponse: true,
+  maxTokens: 0,
+  modelName: 'gpt-4o-mini',
+  modelProvider: false,
+  name: '',
+  tools: [],
+  temperature: 0.2,
+  instructions: dedent(`
+  Given a search object with the following JSON schema: 
+
+  ${minify({
+  "title": "Search",
+  "type": "object",
+  "properties": {
+    "query": {
+      "type": "string",
+      "description": "The search query used"
+    },
+    "results": {
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "url": {
+            "type": "string",
+            "format": "uri",
+            "description": "URL of the web page"
+          },
+          "content": {
+            "type": "string",
+            "description": "Content of the web page at URL in markdown"
+          }
+        },
+        "required": ["url", "content"],
+        "additionalProperties": false
+      },
+      "description": "An array of search results"
+    }
+  },
+  "required": ["query", "results"],
+  "additionalProperties": false
+  })}
+
+  For all search results, silently select all the chunks of text from each result's content that: 
+  - answer the search query directly or provide additional information to support an answer to the search query
+  - contain enough context from the source text to complete the thought expressed by them in the original source text
```

Comment:
Do you think it'd make sense at some point to extract this into a separate, reusable tool?
It seems to not be specific to Google :thinking:

Reply:
Yeah, I think so